File: examples/agentkit/openai/agent.py
"""
OpenAI GPT agent with Scalekit-authenticated Gmail tools.
Mirrors: /agentkit/examples/openai/
Run: python agent.py
"""
import os
import json
import scalekit.client
from openai import OpenAI
from dotenv import load_dotenv
from google.protobuf.json_format import MessageToDict

load_dotenv()

scalekit_client = scalekit.client.ScalekitClient(
    client_id=os.getenv("SCALEKIT_CLIENT_ID"),
    client_secret=os.getenv("SCALEKIT_CLIENT_SECRET"),
    env_url=os.getenv("SCALEKIT_ENV_URL"),
)
actions = scalekit_client.actions
client = OpenAI()

# Connect user to Gmail
response = actions.get_or_create_connected_account(
    connection_name="gmail",
    identifier="docs-example",
)
if response.connected_account.status != "ACTIVE":
    link = actions.get_authorization_link(connection_name="gmail", identifier="docs-example")
    print("Authorize Gmail:", link.link)
    input("Press Enter after authorizing...")

# Fetch and convert tools to OpenAI format
scoped_response, _ = actions.tools.list_scoped_tools(
    identifier="docs-example",
    filter={"connection_names": ["gmail"]},
)
llm_tools = [
    {
        "type": "function",
        "function": {
            "name": MessageToDict(t.tool).get("definition", {}).get("name"),
            "description": MessageToDict(t.tool).get("definition", {}).get("description", ""),
            "parameters": MessageToDict(t.tool).get("definition", {}).get("input_schema", {}),
        },
    }
    for t in scoped_response.tools
]

# Run the agent loop
messages = [{"role": "user", "content": "Fetch my last 5 unread emails and summarize them"}]

while True:
    response = client.chat.completions.create(
        model="claude-sonnet-4-6",
        tools=llm_tools,
        messages=messages,
    )
    message = response.choices[0].message
    if not message.tool_calls:
        print(message.content)
        break

    messages.append(message)
    for tc in message.tool_calls:
        result = actions.execute_tool(
            tool_name=tc.function.name,
            identifier="docs-example",
            tool_input=json.loads(tc.function.arguments),
        )
        messages.append({
            "role": "tool",
            "tool_call_id": tc.id,
            "content": str(result.data),
        })

File: examples/agentkit/openai/agent.ts
/**
 * OpenAI GPT agent with Scalekit-authenticated Gmail tools.
 * Mirrors: /agentkit/examples/openai/
 * Run: tsx openai/agent.ts (from examples/agentkit/)
 */
import { ScalekitClient } from '@scalekit-sdk/node'
import { ConnectorStatus } from '@scalekit-sdk/node/lib/pkg/grpc/scalekit/v1/connected_accounts/connected_accounts_pb'
import OpenAI from 'openai'
import 'dotenv/config'

const scalekit = new ScalekitClient(
  process.env.SCALEKIT_ENV_URL!,
  process.env.SCALEKIT_CLIENT_ID!,
  process.env.SCALEKIT_CLIENT_SECRET!,
)
const openai = new OpenAI()

// Connect user to Gmail
const { connectedAccount } = await scalekit.actions.getOrCreateConnectedAccount({
  connectionName: 'gmail',
  identifier: 'docs-example',
})
if (connectedAccount?.status !== ConnectorStatus.ACTIVE) {
  const { link } = await scalekit.actions.getAuthorizationLink({
    connectionName: 'gmail',
    identifier: 'docs-example',
  })
  console.log('Authorize Gmail:', link)
  process.exit(0)
}

// Fetch and convert tools to OpenAI format
const { tools } = await scalekit.tools.listScopedTools('docs-example', {
  filter: { connectionNames: ['gmail'] },
})
const llmTools: OpenAI.ChatCompletionTool[] = tools.map((t) => ({
  type: 'function',
  function: {
    name: t.tool.definition.name,
    description: t.tool.definition.description,
    parameters: t.tool.definition.input_schema,
  },
}))

// Run the agent loop
const messages: OpenAI.ChatCompletionMessageParam[] = [
  { role: 'user', content: 'Fetch my last 5 unread emails and summarize them' },
]

while (true) {
  const response = await openai.chat.completions.create({
    model: 'claude-sonnet-4-6',
    tools: llmTools,
    messages,
  })
  const message = response.choices[0].message
  if (!message.tool_calls?.length) {
    console.log(message.content)
    break
  }
  messages.push(message)
  for (const tc of message.tool_calls) {
    const result = await scalekit.actions.executeTool({
      toolName: tc.function.name,
      identifier: 'docs-example',
      toolInput: JSON.parse(tc.function.arguments),
    })
    messages.push({ role: 'tool', tool_call_id: tc.id, content: JSON.stringify(result.data) })
  }
}
